File: SUDOKU/C++/src/Sudoku.cpp
//; Name: Aidan Wendorf
//; Email: aidan.j.wendorf@vnaderbilt.edu
//; Date: Sep 24 2023
//; Course: PL
//; Honor statement: I attest that I understand the honor code
//; for this class and have neither
//; given nor received any unauthorized aid on this assignment.
#include "Sudoku.h"
#include <fstream>

/**
 * Constructor for Sudoku class.
 */
Sudoku::Sudoku()
{
}
/**
 * Loads the puzzle object with data from file.
 * Assume input file has format of nine rows of nine digits seperated by a space.
 * @param filename the name of the file to load the puzzle from.
 */
void Sudoku::loadFromFile(std::string filename)
{
    // need to open file
    std::ifstream infile;
    infile.open(filename);
    char nextToken;
    // need to read file into puzzle
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {

            infile >> nextToken;
            puzzle[i][j] = nextToken - '0';
        }
    }
}

/** our format is
  4 3 |   8   | 2 5
6     |       |
      |     1 |   9 4
------+-------+------
9     |     4 |   7
      | 6   8 |
  1   | 2     |     3
------+-------+------
8 2   | 5     |
      |       |     5
  3 4 |   9   | 7 1
 */
/**
 * Prints the puzzle to the output stream.
 * @param os the output stream to print to.
 * @param puzzle the puzzle to print.
 * @return the output stream.
 */
std::ostream& operator<<(std::ostream& os, const Sudoku& puzzle)
{
    for (int i = 0; i < 9; i++) {
        if (i % 3 == 0 && i != 0) {
            os << "------+-------+------" << std::endl;
        }
        for (int j = 0; j < 9; j++) {
            if (j % 3 == 0 && j != 0) {
                os << "| ";
            }

            if (puzzle.puzzle[i][j] == 0) {
                os << " "
                   << " ";
            } else {
                os << puzzle.puzzle[i][j] << " ";
            }
        }
        os << std::endl;
    }
    return os;
}
/**
 * Determines if two puzzles are equal.
 * @param other the puzzle to compare to.
 * @return true if the puzzles are equal, false otherwise.
 */
bool Sudoku::equals(const Sudoku& other) const
{
    for (int i = 0; i < 9; i++) {
        for (int j = 0; i < 9; i++) {
            if (puzzle[i][j] != other.puzzle[i][j]) {
                return false;
            }
        }
    }
    return true;
}
/**
 * Solves the puzzle.
 * @return true if the puzzle was solved, false otherwise.
 */
bool Sudoku::solve()
{
    return recursiveSolve(0, 0);
}
/**
 * Determines if a number is safe to place in the puzzle.
 * @param row the row to check.
 * @param col the column to check.
 * @param num the number to check.
 * @return true if the number is safe to place, false otherwise.
 * @note this is a private helper function for recursiveSolve().
 */
bool Sudoku::isSafe(int row, int col, int num)
{
    // check row
    for (int i = 0; i < 9; i++) {
        if (puzzle[row][i] == num) {
            return false;
        }
    }
    // check col
    for (int i = 0; i < 9; i++) {
        if (puzzle[i][col] == num) {
            return false;
        }
    }
    // check box -> (if the number is already in that 3x3 box)
    int boxRow = row - row % 3;
    int boxCol = col - col % 3;
    for (int i = boxRow; i < boxRow + 3; i++) {
        for (int j = boxCol; j < boxCol + 3; j++) {
            if (puzzle[i][j] == num) {
                return false;
            }
        }
    }
    // if we get here then the number is safe to place
    return true;
}
/**
 * Recursively solves the puzzle.
 * @param row the row to start at.
 * @param col the column to start at.
 * @return true if the puzzle was solved, false otherwise.
 * @note this is a private helper function for solve().
 */
bool Sudoku::recursiveSolve(int row, int col)
{
    // base case
    if (row == 9) {
        return true;
    }
    // if the current spot is not empty
    if (puzzle[row][col] != 0) {
        if (col == 8) {
            return recursiveSolve(row + 1, 0);
        } else {
            return recursiveSolve(row, col + 1);
        }
    }
    for (int i = 1; i <= 9; i++) {
        if (isSafe(row, col, i)) {
            puzzle[row][col] = i;
            if (col == 8) {
                //  if we are at the end of the col
                if (recursiveSolve(row + 1, 0)) {
                    return true;
                }
            } else {
                //  if we are not at the end of the col
                if (recursiveSolve(row, col + 1)) {
                    return true;
                }
            }
        }
    }
    // if we get here then we need to backtrack
    puzzle[row][col] = 0;
    return false;
}

File: SUDOKU/PROLOG/src/sudoku.pl
/*This line is supposed to be a comment. Is it?*/

/*Name: Aidan Wendorf
% Email: aidan.j.wendorf@vanderbilt.edu
% Date: Nov 24 2023
% Course: PL
% Honor statement: I attest that I understand the honor code
% for this class and have neither
% given nor received any unauthorized aid on this assignment.
*/


% Use routines from the Constraint Logic Programming over Finite Domains library.
:- use_module(library('clpfd')).


% The main entry point. Enter "go." at the Prolog prompt.
% 
go :-
    File = 'txt/sudoku-test1.txt',
    start(File).


% The absolute path of the root folder of the project.
% NOTE: MUST BE CHANGED TO THE ABSOLUTE PATH OF ROOT FOLDER OF YOUR PROJECT ON YOUR LOCAL COMPUTER.
%       MAC USERS: USE FORWARD SLASH '/'
%
project_root(Dir) :-
    Dir = '/Users/aidanwendorf/project3-aidanjinn/'.


% To test other than 'sudoku-test1.txt', pass the text file name preceded by 'txt/'.
% For example, enter "start('txt/sudoku-test2.txt')."
%
start(File) :-
    project_root(Dir),                       % Get absolute path to root folder of project.
    atom_concat(Dir, File, PuzzleFile),      % Concatenate path with text file relative location.
    see(PuzzleFile),                         % Open file.
    write(trying_file(PuzzleFile)), nl, nl,  % Display puzzle.
    read(Board),                             % Read board.
    seen,                                    % Close file.
    time(sudoku(Board)),                     % Call solver (with timer).
    pretty_sudo_print(Board), nl.            % Printed solution.


% Print the board to the screen, where each row printed using printsudorow.
%
pretty_sudo_print(Board) :-
    Board = [R1, R2, R3, R4, R5, R6, R7, R8, R9],
    nl, nl,
    printsudorow(R1),
    printsudorow(R2),
    printsudorow(R3),
    write('-------+-------+-------'), nl,
    printsudorow(R4),
    printsudorow(R5),
    printsudorow(R6),
    write('-------+-------+-------'), nl,
    printsudorow(R7),
    printsudorow(R8),
    printsudorow(R9).


% Print row by printing each column.
%
printsudorow(Row) :-
    Row = [C1,C2,C3,C4,C5,C6,C7,C8,C9],
    write(' '),
    write(C1), write(' '),
    write(C2), write(' '),
    write(C3), write(' '), write('|'), write(' '),
    write(C4), write(' '),
    write(C5), write(' '),
    write(C6), write(' '), write('|'), write(' '),
    write(C7), write(' '),
    write(C8), write(' '),
    write(C9), write(' '), nl.

% Base case for the create_blocks/4 predicate.
create_blocks([], [], [], []).

% Recursive case for the create_blocks/4 predicate.
create_blocks([A,B,C|RowTail1],[D,E,F|RowTail2],[G,H,I|RowTail3], [Block|BlocksTail]) :-
    % Create a block by combining the first three elements from each row.
    Block = [A,B,C,D,E,F,G,H,I],
    % Recursively process the remaining elements.
    create_blocks(RowTail1, RowTail2, RowTail3, BlocksTail).

% Helper predicate to divide a 9-element list into three blocks.
divide_into_blocks([A,B,C,D,E,F,G,H,I], Blocks) :-
    % Recursively apply create_blocks/4 to divide the list into three sub-blocks.
    create_blocks(A,B,C,Block1), create_blocks(D,E,F,Block2), create_blocks(G,H,I,Block3),
    % Combine the three sub-blocks into the final list of blocks.
    append([Block1, Block2, Block3], Blocks).

% Main solver containing the rules to solve Sudoku.
sudoku(Puzzle) :-
    % Flatten the Puzzle matrix and constrain each element to be in the range 1..9.
    flatten(Puzzle, FlattenedPuzzle), FlattenedPuzzle ins 1..9,
    % Alias for the Rows matrix.
    Rows = Puzzle,
    % Transpose the Rows matrix to obtain the Columns matrix.
    transpose(Rows, Columns),
    % Divide the Rows matrix into blocks (subgrids).
    divide_into_blocks(Rows, Blocks),
    % Originally used all_different but found using maplist was easier and looked better
    % Ensure that each row has distinct values.
    maplist(all_distinct, Rows),
    % Ensure that each column has distinct values.
    maplist(all_distinct, Columns),
    % Ensure that each block has distinct values.
    maplist(all_distinct, Blocks),
    % Use the label predicate to assign values to variables and find a solution.
    maplist(label, Rows).
